File: io/api/src/logger.ts
/**
 * Structured JSON logs for centralized log systems (one JSON object per line).
 *
 * Canonical schema (see docs/logging.md):
 *   { time, level, msg, component, module, ... }
 *
 * `component` is always the constant "io" since this logger lives inside the
 * io component. The second argument to ioLog/logFromContext is the *module*
 * — the sub-unit within io (e.g. "http", "nats", "transcription").
 */

import type { Context } from 'hono'

export type LogLevel = 'debug' | 'info' | 'warn' | 'error'
export type LogLevelInput = LogLevel | 'warning' | 'fatal' | 'critical'

const COMPONENT = 'io'

const levelRank: Record<LogLevel, number> = {
  debug: 0,
  info: 1,
  warn: 2,
  error: 3,
}

function normalizeLogLevel(raw: string | undefined): LogLevel {
  const v = (raw ?? 'info').toLowerCase()
  if (v === 'debug' || v === 'info' || v === 'warn' || v === 'error') {
    return v
  }
  if (v === 'warning') return 'warn'
  if (v === 'fatal' || v === 'critical') return 'error'
  return 'info'
}

const LOG_LEVEL = normalizeLogLevel(process.env.LOG_LEVEL)

function minLevel(): number {
  return levelRank[LOG_LEVEL]
}

function shouldEmit(level: LogLevel): boolean {
  return levelRank[level] >= minLevel()
}

export type LogFields = Record<string, unknown>

/**
 * Truncate user-supplied text into a debug-safe preview suitable for short
 * metadata fields (titles, reasons, vault progress messages, error codes,
 * voice transcripts).
 *
 * Caps at `maxWords` words AND `maxChars` characters (whichever is hit first)
 * and appends `…` when truncation occurred. Whitespace is collapsed and
 * trimmed so multi-line input renders as a single readable string.
 *
 * For long *conversation* content (user chat messages, agent replies), prefer
 * `previewHeadTail` instead — it keeps the start AND the end so the line
 * remains recognisable when the message is many paragraphs long. See
 * docs/logging.md for the full policy.
 */
export function previewWords(s: string, maxWords = 20, maxChars = 140): string {
  if (!s) return ''
  const flat = s.replace(/\s+/g, ' ').trim()
  if (!flat) return ''
  const words = flat.split(' ')
  let truncated = false
  let out = flat
  if (words.length > maxWords) {
    out = words.slice(0, maxWords).join(' ')
    truncated = true
  }
  if (out.length > maxChars) {
    out = out.slice(0, maxChars).trimEnd()
    truncated = true
  }
  return truncated ? `${out}…` : out
}

/**
 * Build a debug-safe head+tail preview suitable for long conversation
 * messages — typically `content_preview` (user → agent) and `result_preview`
 * (agent → user).
 *
 * Format: `<first headWords words> [..N chars..] <last tailWords words>`,
 * where `N` is the number of characters omitted from the middle. When the
 * string is short enough that head+tail would already cover the whole thing,
 * the original (whitespace-collapsed) value is returned unchanged.
 *
 * The motivation is debugger UX: a user might paste a long document and put
 * the actual question on the last line; a beginning-only preview hides it.
 * Head+tail makes the same message recognisable across IO, orchestrator, and
 * agent logs.
 */
export function previewHeadTail(s: string, headWords = 15, tailWords = 10): string {
  if (!s) return ''
  const flat = s.replace(/\s+/g, ' ').trim()
  if (!flat) return ''
  const words = flat.split(' ')
  if (words.length <= headWords + tailWords) {
    return flat
  }
  const head = words.slice(0, headWords).join(' ')
  const tail = words.slice(words.length - tailWords).join(' ')
  const omitted = flat.length - head.length - tail.length
  if (omitted <= 0) return flat
  return `${head} [..${omitted} chars..] ${tail}`
}

/**
 * Emit one structured log line (no request context).
 *
 * @param level   canonical level or alias (warning|fatal|critical)
 * @param module  sub-unit name within the io component (e.g. "http", "nats")
 * @param msg     short event name or sentence
 * @param fields  additional safe metadata
 */
export function ioLog(
  level: LogLevelInput,
  module: string,
  msg: string,
  fields: LogFields = {},
): void {
  const normalizedLevel = normalizeLogLevel(level)
  if (!shouldEmit(normalizedLevel)) return
  const rec: Record<string, unknown> = {
    time: new Date().toISOString(),
    level: normalizedLevel.toUpperCase(),
    component: COMPONENT,
    module,
    msg,
    ...fields,
  }
  const line = JSON.stringify(rec)
  if (normalizedLevel === 'error') console.error(line)
  else console.log(line)
}

/**
 * Structured log with trace_id, task_id, and conversation_id pulled from Hono
 * context. trace middleware sets traceId; route handlers set taskId and
 * conversationId via c.set(...) at the top of each handler. Explicit fields
 * passed in `fields` win over context values, so a route can override on a
 * single line.
 */
export function logFromContext(
  c: Context,
  level: LogLevelInput,
  module: string,
  msg: string,
  fields: LogFields = {},
): void {
  const traceId = c.get('traceId') as string | undefined
  const taskId = c.get('taskId') as string | undefined
  const conversationId = c.get('conversationId') as string | undefined
  const merged: LogFields = { ...fields }
  if (taskId !== undefined && merged.task_id === undefined) merged.task_id = taskId
  if (conversationId !== undefined && merged.conversation_id === undefined) merged.conversation_id = conversationId
  logFromContextWithTrace(traceId, level, module, msg, merged)
}

export function logFromContextWithTrace(
  traceId: string | undefined,
  level: LogLevelInput,
  module: string,
  msg: string,
  fields: LogFields = {},
): void {
  const merged = traceId ? { ...fields, trace_id: traceId } : { ...fields }
  ioLog(level, module, msg, merged)
}

File: orchestrator/internal/observability/logger.go
package observability

import (
	"context"
	"log/slog"
	"os"
	"strconv"
	"strings"
	"unicode/utf8"
)

// nodeID is the node identity set at startup via InitLogger.
var nodeID string

// defaultLogger is the global structured logger, initialized by InitLogger.
var defaultLogger *slog.Logger

func init() {
	// Provide a safe default so tests that don't call InitLogger still get a valid logger.
	defaultLogger = slog.New(slog.NewJSONHandler(os.Stdout, &slog.HandlerOptions{
		Level: slog.LevelInfo,
	}))
}

// InitLogger sets up the global structured logger.
// Must be called once from main.go before any log output.
//
//   - level:  "debug" | "info" | "warn" | "warning" | "error" | "fatal" | "critical" (default: "info")
//   - format: accepted for backward compatibility; logs are always JSON
//   - node:   NODE_ID value for the node_id field
func InitLogger(level, format, node string) {
	_ = format
	nodeID = node

	var lvl slog.Level
	switch strings.ToLower(level) {
	case "debug":
		lvl = slog.LevelDebug
	case "warn", "warning":
		lvl = slog.LevelWarn
	case "error", "fatal", "critical":
		lvl = slog.LevelError
	default:
		lvl = slog.LevelInfo
	}

	opts := &slog.HandlerOptions{Level: lvl}
	defaultLogger = slog.New(slog.NewJSONHandler(os.Stdout, opts))
}

// LoggerWithModule returns the default orchestrator logger with the canonical
// component=orchestrator attribute and the supplied module name. Prefer
// LogFromContext when context IDs are available.
func LoggerWithModule(module string) *slog.Logger {
	attrs := []any{"component", "orchestrator", "module", module}
	if nodeID != "" {
		attrs = append(attrs, "node_id", nodeID)
	}
	return defaultLogger.With(attrs...)
}

// LoggerWithComponent is a deprecated alias for LoggerWithModule kept so older
// call sites compile. Its argument is the module name within the orchestrator
// component, despite the legacy name.
//
// Deprecated: use LoggerWithModule.
func LoggerWithComponent(module string) *slog.Logger {
	return LoggerWithModule(module)
}

// LogFromContext returns a *slog.Logger pre-populated with all IDs present in ctx.
//
// Required fields on every line (EDD §15.1):
//   - component       = "orchestrator"
//   - node_id         (from InitLogger)
//   - trace_id        (from context, if set)
//   - task_id         (from context, if set)
//   - conversation_id (from context, if set)
//   - plan_id         (from context, if set)
//   - subtask_id      (from context, if set)
//   - module          (from context, if set)
//
// FORBIDDEN log content (EDD §15.1):
//   - raw user input (payload.raw_input)
//   - credential values
//   - task result payloads
//   - planner output
func LogFromContext(ctx context.Context) *slog.Logger {
	attrs := []any{}
	if v := TraceIDFrom(ctx); v != "" {
		attrs = append(attrs, "trace_id", v)
	}
	if v := TaskIDFrom(ctx); v != "" {
		attrs = append(attrs, "task_id", v)
	}
	if v := ConversationIDFrom(ctx); v != "" {
		attrs = append(attrs, "conversation_id", v)
	}
	if v := PlanIDFrom(ctx); v != "" {
		attrs = append(attrs, "plan_id", v)
	}
	if v := SubtaskIDFrom(ctx); v != "" {
		attrs = append(attrs, "subtask_id", v)
	}
	if v := ModuleFrom(ctx); v != "" {
		attrs = append(attrs, "module", v)
	}
	// Note: when the context carries no module, we still emit component=orchestrator
	// so every line has the canonical component label even if the caller has not
	// chosen a module yet.
	base := []any{"component", "orchestrator"}
	if nodeID != "" {
		base = append(base, "node_id", nodeID)
	}
	return defaultLogger.With(base...).With(attrs...)
}

// PreviewWords truncates user-supplied text into a debug-safe preview suitable
// for short metadata fields (titles, reasons, error codes, progress messages,
// voice transcripts).
//
// Caps at maxWords words AND maxChars characters (whichever is hit first) and
// appends "…" when truncation occurred. Whitespace is collapsed and trimmed so
// multi-line input renders as a single readable string.
//
// For long *conversation* content (user chat messages, agent replies), prefer
// PreviewHeadTail — it keeps the start AND the end so the line remains
// recognisable when the message is many paragraphs long. See docs/logging.md
// for the full policy.
func PreviewWords(s string, maxWords, maxChars int) string {
	if maxWords <= 0 {
		maxWords = 20
	}
	if maxChars <= 0 {
		maxChars = 140
	}
	flat := strings.Join(strings.Fields(s), " ")
	if flat == "" {
		return ""
	}
	words := strings.Split(flat, " ")
	truncated := false
	out := flat
	if len(words) > maxWords {
		out = strings.Join(words[:maxWords], " ")
		truncated = true
	}
	if utf8.RuneCountInString(out) > maxChars {
		runes := []rune(out)
		out = strings.TrimRight(string(runes[:maxChars]), " ")
		truncated = true
	}
	if truncated {
		return out + "…"
	}
	return out
}

// PreviewHeadTail builds a debug-safe head+tail preview suitable for long
// conversation messages — typically content_preview (user → agent) and
// result_preview (agent → user).
//
// Format: "<first headWords words> [..N chars..] <last tailWords words>",
// where N is the number of characters omitted from the middle. When the
// string is short enough that head+tail would already cover the whole value,
// the original (whitespace-collapsed) string is returned unchanged.
//
// The motivation is debugger UX: a user might paste a long document and put
// the actual question on the last line; a beginning-only preview would hide
// it. Head+tail also makes the same message recognisable across io,
// orchestrator, and agent logs.
func PreviewHeadTail(s string, headWords, tailWords int) string {
	if headWords <= 0 {
		headWords = 15
	}
	if tailWords <= 0 {
		tailWords = 10
	}
	flat := strings.Join(strings.Fields(s), " ")
	if flat == "" {
		return ""
	}
	words := strings.Split(flat, " ")
	if len(words) <= headWords+tailWords {
		return flat
	}
	head := strings.Join(words[:headWords], " ")
	tail := strings.Join(words[len(words)-tailWords:], " ")
	omitted := utf8.RuneCountInString(flat) - utf8.RuneCountInString(head) - utf8.RuneCountInString(tail)
	if omitted <= 0 {
		return flat
	}
	return head + " [.." + strconv.Itoa(omitted) + " chars..] " + tail
}
